lib: refactor exposed internal modules

aduh95 opened the pull request:
Move the implementation of internal modules onto the `lib/internal/` folder, and add aliases to keep the backward compatibility in case user code relies on them being part of the public API.

It also replace usage of those modules into the `test/` folder to their documented equivalent APIs when possible.

The idea behind this PR is to ease the process of deprecation if we ever decided that one of those modules should be deprecated at some point, and make the `lib/` folder structure more consistent.

Refs: https://github.com/nodejs/node/pull/33326
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

aduh95 commented:
@jasnell To be clear, this does not remove any public module, nor does it break any code that use them. The only change is that core would use `require('internal/http/common')` instead of `require('_http_common')` and so on, but both are still valid and expose the same API. Sorry if the OP was confusing, I'll try to rephrase it.

jasnell commented:
Ah right... I saw the deletes and hadn't scrolled down to the bottom to see the re-exports ;-) ... either way, I don't think we should actually do this. These have been this way for quite some time and these aren't likely to be able to go away at all. I'd prefer just to keep them in place how they are currently.

benjamingr commented:
Generally this looks good to me - the only part I'm not sure about is that now that the tests require the external public modules we don't have tests in place to check that these imports still work - so I'd expect a test for that :]

I'm also +0 on this change since I'm used to where these files are - but not blocking.

joyeecheung commented:
Can you try making git recognize the files are moved instead of being added/removed? Otherwise I feel that this will make backports much more difficult and is doing more harm than good..

devsnek commented:
@joyeecheung i don't think it can detect it if the original file still exists and is modified in the same commit

BridgeAR commented:
While not thinking strongly about it, it does seem like a good idea to move those in `internal`. It always felt weird to have the underscored file in `lib`.

I do not think that backporting is actually an issue as long as these are backported right away. @nodejs/releasers what do you think? I guess this mainly applies to us.

aduh95 commented:
> Can you try making git recognize the files are moved instead of being added/removed? Otherwise I feel that this will make backports much more difficult and is doing more harm than good..

I tried to split it in two commits, but I'm not sure if that's enough...

> we don't have tests in place to check that these imports still work

That's now fixed, I have added a test PTAL.

> I'm also +0 on this change since I'm used to where these files are

I clearly don't have as much experience with the code base as you, so for me it's just weird to have those modules meant to be internal here. IMHO they should be moved for the sake of consistency, but no hard feeling if we decide to not land this.

aduh95 commented:
@jasnell Are you still -1 on this? If you are, I guess we can close it for now. Otherwise, I'd be happy to work on a rebase to resolve the merge conflict.

jasnell commented:
Yeah, I'm sorry but I am -1 on it. I just don't see the benefit of moving these in. I appreciate the work put into this tho

aduh95 commented:
Closing as Won't fix.
